crates/fuel-core/src/combined_database.rs

```diff
@@ -312,17 +312,46 @@ impl CombinedDatabase {
         Ok(())
     }
 
+    /// This function is fundamentally different from `rollback_to` in that it
+    /// will rollback the off-chain/gas-price databases if they are ahead of the
+    /// on-chain database. If they don't have a height or are behind the on-chain
+    /// we leave it to the caller to decide how to bring them up to date.
+    /// We don't rollback the on-chain database as it is the source of truth.
+    /// The target height of the rollback is the latest height of the on-chain database.
     pub fn sync_aux_db_heights<S>(&self, shutdown_listener: &mut S) -> anyhow::Result<()>
     where
         S: ShutdownListener,
     {
-        if let Some(on_chain_height) = self.on_chain().latest_height_from_metadata()? {
-            // todo(https://github.com/FuelLabs/fuel-core/issues/2239): This is a temporary fix
-            let res = self.rollback_to(on_chain_height, shutdown_listener);
-            if res.is_err() {
-                tracing::warn!("Failed to rollback auxiliary databases to on-chain database height: {:?}", res);
+        while !shutdown_listener.is_cancelled() {
+            let on_chain_height = match self.on_chain().latest_height_from_metadata()? {
+                Some(height) => height,
+                None => break, // Exit loop if on-chain height is None
+            };
+
+            let off_chain_height = self.off_chain().latest_height_from_metadata()?;
+            let gas_price_height = self.gas_price().latest_height_from_metadata()?;
+
+            // Handle off-chain rollback if necessary
+            if let Some(off_height) = off_chain_height {
+                if off_height > on_chain_height {
+                    self.off_chain().rollback_last_block()?;
```

Comment:
What if the `rollback_last_block` fails after rolling back a couple blocks. Are we okay with mutating the state without correcting it?

Reply:
good question ~ if we include this refactor, we will also have to ensure the existing function on `CombinedDatabase::rollback_to` will have to be refactored - 
https://github.com/FuelLabs/fuel-core/blob/103ecf6f49db2c04b3c4f88f3d87d8f6ff2d47c2/crates/fuel-core/src/combined_database.rs#L234-L313

a simple way would perhaps be to use a `WriteTransaction` and then commit once done.

Reply:
how does this look @MitchTurner? - 54457c21e3

Reply:
eh, looks like the internal function doesn't actually rollback atomically. we'll have to keep the loop. it rolls back the blocks one-by-one, which makes the `rollback_to` quite misleading as a function name in https://github.com/FuelLabs/fuel-core/blob/b1f2bf445a2e9a270f1c0032f20c680d8e15d251/crates/fuel-core/src/state.rs#L61

cc: @xgreenx for confirmation

Reply:
I'm going to revert [54457c2](https://github.com/FuelLabs/fuel-core/pull/2272/commits/54457c21e3c2cac4a00a4f3c5d01fe07ce4a0ef2) for now

Reply:
I think we are okay with mutating state since during next launch we continue to rollback the state. It doesn't live as in corrupted state.